src/bin/languageserver/mod.rs

```diff
@@ -15,42 +17,92 @@ use solang_parser::pt;
-        MarkedString, MessageType, OneOf, Position, Range, ServerCapabilities,
-        SignatureHelpOptions, TextDocumentContentChangeEvent, TextDocumentSyncCapability,
-        TextDocumentSyncKind, Url, WorkspaceFoldersServerCapabilities, WorkspaceServerCapabilities,
+        ExecuteCommandOptions, ExecuteCommandParams, GotoDefinitionParams, GotoDefinitionResponse,
+        Hover, HoverContents, HoverParams, HoverProviderCapability, InitializeParams,
+        InitializeResult, InitializedParams, Location, MarkedString, MessageType, OneOf, Position,
+        Range, ServerCapabilities, SignatureHelpOptions, TextDocumentContentChangeEvent,
+        TextDocumentSyncCapability, TextDocumentSyncKind, Url, WorkspaceFoldersServerCapabilities,
+        WorkspaceServerCapabilities,
     },
     Client, LanguageServer, LspService, Server,
 };
 
 use crate::cli::{target_arg, LanguageServerCommand};
 
-struct Hovers {
-    file: ast::File,
-    hovers: Lapper<usize, String>,
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum DefinitionType {
+    Function(usize),
+    Variable(usize),
+    NonLocalVariable(Option<usize>, usize),
+    Struct(usize),
+    Field(Type, usize),
+    Enum(usize),
+    Variant(usize, usize),
+    Contract(usize),
+    Event(usize),
+    UserType(usize),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct DefinitionIndex {
+    def_path: PathBuf,
+    def_type: DefinitionType,
 }
 
+/// Stores locations of definitions of functions, contracts, structs etc.
+type Definitions = HashMap<DefinitionIndex, Range>;
+/// Stores strings shown on hover
 type HoverEntry = Interval<usize, String>;
+/// Stores locations of function calls, uses of structs, contracts etc.
+type ReferenceEntry = Interval<usize, DefinitionIndex>;
+
+struct FileCache {
+    file: ast::File,
+    hovers: Lapper<usize, String>,
+    references: Lapper<usize, DefinitionIndex>,
+}
 
 /// Stores information used by language server for every opened file
 struct Files {
-    hovers: HashMap<PathBuf, Hovers>,
+    caches: HashMap<PathBuf, FileCache>,
     text_buffers: HashMap<PathBuf, String>,
 }
 
+// The language server currently stores some of the data grouped by the file to which the data belongs (Files struct).
+// Other data (Definitions) is not grouped by file due to problems faced during cleanup,
+// but is stored as a "global" field which is common to all files.
+//
+// When the file is closed. This triggers the `did_close` handler function
+// where we remove the entry corresponding to the closed file from Files::cache.
+// If the definitions are part of `FileCache`, they are also removed with the rest of `FileCache`
+// But there can be live references in other files whose definitions are defined in the closed file.
+//
+// Files from multiple namespaces can be open at any time in VSCode.
+// But compiler currently works on the granularity of a a namespace,
+// i.e, all the analyses + code generated is for the whole namespace.
```

Comment:
You said that the compiler works on the granularity (i.e. individual pieces), but then you explain that the analysis and code generated is for the whole namespace. Did you mean entirety instead of granularity?

Reply:
I was going for the following meaning of "granularity" here. 
> the scale or level of detail in a set of data.

Reply:
That makes sense. Perhaps you need to rephrase what you wrote after the `i.e` then. It is a little confusing.

Reply:
Hello!
I decided to remove the line as it didn't add any more information and was redundant. Thanks!